### DISCUSSION TITLE
Run Xamarin UI Tests for iOS and Android under Github Actions

### DISCUSSION THREAD

**Discussion by StuartFerguson** [community/community#25069] [Actions]
<p>I have written a mobile application using Xamarin Forms under Visual Studio 2019, this targets both Android and iOS.</p>
<p>As part of my development I have written some Xamarin.UITest based End to End tests using Specflow and NUnit.</p>
<p>I am now trying to get these to run as part of my CI workflow and I am having a bit of difficulty.</p>
<p>Mainly this is around the area of actually getting the Android and iOS emulators to fire up so the tests can be executed.</p>
<p>Is there something that I can add to my workflow that will start up the Android Emulator with a specified andoid version that my tests can be run against and then anotehr command to close down the emulator after my tests have finished running.</p>
<p>So far I have been running the whole workflow on a Windows OS as the code is Net Framework.</p>
<p>Ideally some way of running iOS based tests in a CI workflow would be helpful.</p>
<p>Regards,</p>
<p>Stuart</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/stuartferguson">@stuartferguson</a> ,</p>
<p>Is it resolved by action ‘malinskiy/action-android/emulator-run-cmd’ for your Xamarin UI test on macOS? To kill the emulator and finish the action, you can also try similar action ‘<a href="https://github.com/ReactiveCircus/android-emulator-runner" rel="nofollow noopener">ReactiveCircus/android-emulator-runner</a>’. Thanks.</p>

**Comment by StuartFerguson** [comment]
<p>Hi <a class="mention" href="/weide-zhou">@weide-zhou</a></p>
<p>Thanks for the reply.</p>
<p>I managed to get it working with the malinskiy/action-android/emulator-run-cmd action.</p>
<p>Regards,</p>
<p>Stuart</p>

**Comment by StuartFerguson** [comment]
<p><a class="mention" href="/weide-zhou">@weide-zhou</a>  another question on the above.</p>
<p>As part of my UI tests for Android and iOS I need to spin up a docker container and then communicate with that from my mobile application.</p>
<p>As part of the test background, I am posting data to the docker container from the hosting machine.  This scenario works fine on my windows development environment, however when running on the macos Github Agent these setup POST’s are getting as connection refused from the docker container address:port.</p>
<p>Are you able to offer any advice on how I can get this working?</p>
<p>The workflow I am using is</p>
<p><a href="https://github.com/StuartFerguson/MobileApplication/blob/master/.github/workflows/pullrequest.yml" title="Workflow" rel="nofollow noopener">https://github.com/StuartFerguson/MobileApplication/blob/master/.github/workflows/pullrequest.yml</a> </p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/stuartferguson">@stuartferguson</a> ,</p>
<p>Much appreciate your mark!</p>
<p>Regarding the new problem, I didn’t find the exact error but i notice the workflow passed on branch ‘task/<span class="hashtag">#3_loginscreen</span>’, is the problem resolved?</p>
<p>Thanks.</p>
